pages/partials/attachFilePanel.ejs

```diff
@@ -0,0 +1,100 @@
+            <i class="far fa-trash-alt"></i>
+          </a>
+          <% } %>
+        </td>
+        <% } %>
+      </tr>
+      <% }); %>
+    <% } %>
+    </tbody>
+  </table>
+
+  <div class="card-footer">
+    <% if (!assessment_instance.open) { %>
+    <p class="small mb-0">
+      Attachments can't be added or deleted because the assessment is closed.
+    </p>
+    <% } else { /* assessment_instance is open */ %>
+    <!-- UPLOAD FILE -->
+    <button class="btn btn-xs btn-secondary" type="button" data-toggle="collapse" data-target="#attachFileCollapse" aria-expanded="false" aria-controls="attachFileCollapse">
+      Attach a file <i class="far fa-caret-square-down"></i>
+    </button>
+    <div class="collapse" id="attachFileCollapse">
+      <form class="attach-file-form" name="attach-file-form" method="POST" enctype="multipart/form-data" class="mb-3">
+        <p class="small mt-3">
+          Attached files will be saved here for your reference.
+        </p>
+        <div class="form-group">
+          <div class="custom-file">
+            <input type="file" name="file" class="custom-file-input" id="attachFileInput">
+            <label class="custom-file-label" for="attachFileInput">Choose file</label>
+            <small id="emailHelp" class="form-text text-muted">Max file size: <%= config.fileUploadMaxBytesFormatted %></small>
+          </div>
+        </div>
+        <div class="d-flex justify-content-end">
+          <div class="form-group mb-0">
+            <input type="hidden" name="__action" value="attach_file">
+            <% if (typeof variant != 'undefined') { %>
+            <input type="hidden" name="__variant_id" value="<%= variant.id %>">
+            <% } %>
+            <input type="hidden" name="__csrf_token" value="<%= __csrf_token %>">
+            <button type="submit" class="btn btn-primary">Attach file</button>
+          </div>
+        </div>
+      </form>
+    </div>
+
+    <!-- UPLOAD TEXT -->
+    <br>
+    <button class="btn btn-xs btn-secondary" type="button" data-toggle="collapse" data-target="#attachTextCollapse" aria-expanded="false" aria-controls="attachTextCollapse">
+      Attach text <i class="far fa-caret-square-down"></i>
+    </button>
+    <div class="collapse" id="attachTextCollapse">
+      <form method="POST" class="attach-text-form">
+        <p class="small mt-3">
+          Attached text will be saved here for your reference.
+        </p>
+        <input class="form-control" name="filename" value="notes.txt">
+        <div class="form-group">
+          <textarea class="form-control" rows="5" name="contents" placeholder="Type or paste text here"></textarea>
+        </div>
+        <% if (typeof variant != 'undefined') { %>
+        <input type="hidden" name="__variant_id" value="<%= variant.id %>">
+        <% } %>
+        <input type="hidden" name="__csrf_token" value="<%= __csrf_token %>">
+        <div class="form-group text-right">
```

Comment:
Nit: I'd prefer a left-aligned button. For a single-column form, the eye wants to naturally keep moving down the page after the inputs, and putting the button on the right makes it slightly harder to notice IMO.

Reply:
good point, done